src/utils/build.ts

```diff
@@ -0,0 +1,143 @@
+		} else {
+			let hubCluster = hubClusters[0];
+			// create kind clusters
+			progressReporter({increment: 0 , message: `creating ${clusters.length} kind clusters`});
+			let clusterPromises = clusters.map(cluster =>
+				createKindCluster(cluster)
+					.then(stdout => console.log(stdout))
+					.catch(stderr => console.log(stderr))
+			);
+			Promise.all(clusterPromises)
+				.then(() => {
+					console.debug('created kind clusters successfully');
+					progressReporter({increment: 20 , message: `initializing the Hub cluster named ${hubCluster.name}`});
+					// initializing the hub cluster
+					initializeHubCluster(hubCluster)
+						.then((joinCmd: string) => {
+							console.debug('initialized hub cluster successfully');
+							progressReporter({increment: 20 , message: 'issuing join requests for the managed clusters'});
+							// issue join requests from the managed clusters to the hub
+							managedClusters.reduce(
+								(previousPromise, currentPromise) =>
+									previousPromise.then(() =>
+										issueJoinRequest(currentPromise, joinCmd)), // TODO: loosing the stdout/stderr here
+									Promise.resolve('initial value')
+							)
+							.then(() => {
+								console.debug('issued join requests successfully');
+								progressReporter({increment: 20 , message: 'accepting the managed clusters join request from the hub cluster'});
+								// accept the issued join commands by the managed clusters from the hub cluster
+								acceptJoinRequests(hubCluster, managedClusters)
+									.then(() => {
+										console.debug('accepted join requests successfully');
+										progressReporter({increment: 100 , message: 'successfully created your development environment, have fun'});
+										// TODO: display the user with information about the created clusters: names/contexts
+										resolve('OCM extension, successfully created your development environment, have fun');
+									})
+									.catch((stderr) => {
+										console.debug('failed to accept join requests');
+										console.error(stderr);
+										progressReporter({increment: 100 , message: 'failed to accept the join request made from the managed clusters'});
+										reject('OCM extension, failed to accept the join request made from the managed clusters');
+									});
+							})
+							.catch(() => {
+								console.debug('failed to issue join requests');
+								progressReporter({increment: 100 , message: 'failed to make join request for the managed clusters'});
+								reject('OCM extension, failed to make join request for the managed clusters');
+							});
+						})
+						.catch((stderr: string) => {
+							console.debug('failed initializing the hub cluster');
+							console.error(stderr);
+							progressReporter({increment: 100 , message: 'failed to initialize the hub cluster'});
+							reject('OCM extension, failed to initialize the hub cluster');
+						});
+				})
```

Comment:
this is less clean than using await.

Reply:
I think so.
before, after each _await_ statement, we needed to _sleep_,
using this callbacks technique, and we sleep no more.
:-)

Reply:
also, with the awaits statements,
we dropped granular control of the displayed message.

we didn't handle rejected promises,
we just let them all drop into the catch clause of the try-catch statement.

now, using the catch callbacks, we handle rejections better at the time they arrive,
we have full control over what is being displayed to the user, and what is being done after the rejection.

if we were to try and do this with await statements,
I'm guessing it wouldn't have looked all that well with a million try-catch/switch/if statements.
🤔 

IMO
the use of the await statements was cool and pretty,
but, as it was, it didn't get half the work done as the callback technique.
😃

Reply:
but we could handle the rejection of the promise. 
e.g 
acceptJoinRequests rejection promise could also take care of  

```
console.debug('failed to accept join requests');
console.error(stderr);
progressReporter({increment: 100 , message: 'failed to accept the join request made from the managed clusters'});
reject('OCM extension, failed to accept the join request made from the managed clusters');
```

Reply:
that means that the utility functions _initializeHubCluster_, _acceptJoinRequests_, and _issueJoinRequest_,
should now also take the _progressReporter_ and the _reject_ callbacks.

granted, this will make the _buildLocalEnv_ exported function much smaller,
but I'm not so sure it will make it more readable, it might be the other way around.